db/db_impl/db_impl_write.cc

```diff
@@ -1195,9 +1205,13 @@ IOStatus DBImpl::WriteToWAL(const WriteBatch& merged_batch,
     *log_used = logfile_number_;
   }
   total_log_size_ += log_entry.size();
-  // TODO(myabandeh): it might be unsafe to access alive_log_files_.back() here
-  // since alive_log_files_ might be modified concurrently
-  alive_log_files_.back().AddSize(log_entry.size());
+  if (with_db_mutex || with_log_mutex) {
+    alive_log_files_.back().AddSize(log_entry.size());
+  } else {
+    assert(!two_write_queues_);
+    InstrumentedMutexLock wl(&mutex_);
```

Comment:
This seems likely to be expensive on a hot path (every WAL write). Can we save `alive_log_files_.back()` to another field each time we `push_back()`? You could probably even create a custom wrapper class that handles that, though that's arguably more complex than this direct approach.

Reply:
The idea of caching sounds good, and it is worth implementing for non-two-write-queues case. For two-write-queues, the `alive_log_files_` is already protected by `log_mutex_`.

If we cache `alive_log_files_.back()` by value, then we still need to sync the cached copy with `alive_log_files_.back()` at a later point when we need to acquire the mutex. We cannot directly cache `&alive_log_files_.back()` by address because it can be copied when deque grows and shrinks.

If we do caching, then I would vote for changing the type of `alive_log_files_` from `std::deque<LogFiileNumberSize>` to `std:deque<std::unique_ptr<LogFileNumberSize>>`, caching the last element via a member of type `LogFileNumberSize* alive_log_files_tail_copy_`

Reply:
Great, then we can avoid calling `back()` here